### ISSUE TITLE
Quarkus Test doesn't have compatibility with PIT

### ISSUE THREAD

**Issue by gdcmarinho** [hcoles/pitest#1021]
I've been creating a back-end service with Quarkus (tests in JUnit/PIT) and I'm not having a good experience with mutation testing (specifically with the first mutant I have to kill). The mutant is: `replaced return value with null for org/company/controller/UserController::getUserInformations → SURVIVED`.

**Controller**
```java
@POST
@Consumes(MediaType.APPLICATION_JSON)
@Produces(MediaType.APPLICATION_JSON)
public UserInformations getUserInformations(String userId) {
    return service.getUserInformations(userId);
}
```

I have some unit tests to this controller method. Thinking about the null mutant, I just added a Assertions.assertNotNull(userObject) in one of these tests. But, this isn't working at all. So, how can I kill the mutant?

By the way, I've already ensured that the tests in question are being executed in the footer of the execution report:

<img width="1007" alt="footer screenshot" src="https://user-images.githubusercontent.com/34725877/169046210-4f816460-ebe2-4b7d-bd40-38f749a523da.png">

**Comment by hcoles** [comment]
If you can post a minimal project that reproduces the issue, I can take a look.

**Comment by gdcmarinho** [comment]
I sent an invite to you. The repo is [located here](https://github.com/gdcmarinho/quarkus-mutation-testing).
<img width="1104" alt="invite project" src="https://user-images.githubusercontent.com/34725877/169052266-d4e00478-62e1-4802-bf01-523c2212379f.png">

**Comment by hcoles** [comment]
Having taken a brief look, the issue is that your tests do not run the code normally in the jvm.

I've not yet got my head around what `@QuarkusTest` is doing that is preventing pitest inserting mutations, but it is clear that they are not active in the bytecode it is running.

`@QuarkusTest` seems to be quite heavy weight, launching the quarkus stack. There is no need to do this for your tests. They use only mocks and direct method calls, so could be implemented without the heavyweight runner.

Hacking the tests to not use `@QuarkusTest` shaves a couple of seconds off the normal test execution time (which is a lot for only 2 tests), and means that pitest can kill the mutants.

I'll look deeper into what `QuarkusTest` is doing when I have chance. It may be possible to support it, but even if pitest does support it, I would still not suggest using it without good reason. I'd save it for integration tests that actually need it.

Using plain old mockito tests, pitest analysed the project in 6 seconds. It took 1 minute 7 seconds using the quarkus tests as they run so much slower.

**Comment by hcoles** [comment]
Having blundered about a little in the code for `QuarkusTestExtension` it doesn't look to be doing anything too unusual/problematic (like launching jvms or instrumenting code). Issue is likely around the fact that it resets the context class loader. If so, it ought to be fixable (with the earlier performance warning).

**Comment by gdcmarinho** [comment]
How was the test class working without quarkus test? I couldn't reproduce it here. :(

**Comment by hcoles** [comment]
You need to change  your controller and service classes to use constructor injection so they can be constructed normally with new. Then write tests as you would for any other class.

**Comment by hcoles** [comment]
Got a bit further into what is happening here. Quarkus test looks to be running the classes in a non delegating class loaders. Pitest is mutating the class, but only in the context class loader set before the quarkus tests runs. The test therefore does not see the mutation.

I have a partially working fix (works for quarkus, but breaks things for gwt mockito) that will hopefully become a fully working fix in the next release.

To give you an idea of the impact of using QuarkusTest. Using normal tests, the project is analysed in less than 10 seconds. It takes 37 seconds using quarkus tests when the fix is in place and they are able to kill mutants. This would quickly become a problem in larger projects.

**Comment by hcoles** [comment]
@gdcmarinho https://github.com/hcoles/pitest/releases/tag/1.8.0

**Comment by gdcmarinho** [comment]
Amazing work, @hcoles! Thank you so much for the support.

**Comment by hcoles** [comment]
@gdcmarinho No problem, it was an interesting one. I suspect there are other test runners out there that have a similar problem that hasn't been reported. Unfortunately I couldn't find a clean general soloution to this, so if any more are identified, the code will have to be updated to recognise them and apply the fix.

Closing as fixed in 1.8.0.
